sv.c

```diff
@@ -4771,6 +4776,204 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
         SvTAINT(dsv);
 }
 
+/* A helper for newSVsv_flags, either directly or indirectly via S_sv_freshcopy_flags.
+ * It is more efficient for their use cases than calling Perl_sv_setsv_flags. */
+PERL_STATIC_INLINE void
+S_sv_freshcopy_POK(pTHX_ SV* dsv, SV* ssv, const I32 flags)
+{
+            const U32 sflags = SvFLAGS(ssv);
+            /* This logic MUST mirror what is in Perl_sv_setsv_flags.
+             * Please refer to the comments there for the explanation
+             * of what is going on below.
+             */
+            assert(!(SvFLAGS(dsv) & SVf_BREAK));
+            assert(!SvIsCOW_static(ssv)); /* S_sv_freshcopy_flags handles this. */
+            assert(!SvTEMP(dsv));
+
+            /* SvLEN, SvCUR, SvPVX are all uninitialized at this point and could contain anything */
+            /* S_SvPV_shared_hkey_or_CoWable() needs an accurate SvLEN(dsv) */
+            SvLEN_set(dsv, 0);
+
+            const STRLEN cur = SvCUR(ssv);
+            const STRLEN len = SvLEN(ssv);
+
+            if (!(flags & SV_NOSTEAL) && S_SvPV_can_swipe_buf(ssv, sflags, cur, len) ) {
+                /* Passes the swipe test.  */
+                SvPV_set(dsv, SvPVX_mutable(ssv));
+                SvLEN_set(dsv, len);
+                SvCUR_set(dsv, cur);
+
+                assert(!SvOOK(ssv)); /* According to S_SvPV_can_swipe_buf() */
+                /* This is (void)SvOK_off(ssv); but without the superfluous SvOOK_off(ssv)) */
+                SvFLAGS(ssv) &= ~(SVf_OK|SVf_IVisUV|SVf_UTF8);
+
+                SvPV_set(ssv, NULL);
+                SvLEN_set(ssv, 0);
+                SvCUR_set(ssv, 0);
+                SvTEMP_off(ssv);
+            } else if ((flags & SV_COW_SHARED_HASH_KEYS) &&
+                S_SvPV_shared_hkey_or_CoWable(ssv, dsv, sflags, cur, len)
+            ) {
+                /* Either it's a shared hash key, or it's suitable for
+                   copy-on-write.  */
+#ifdef DEBUGGING
+                if (DEBUG_C_TEST) {
+                    PerlIO_printf(Perl_debug_log, "Copy on write: ssv --> dsv\n");
+                    sv_dump(ssv);
+                    sv_dump(dsv);
+                }
+#endif
+#ifdef PERL_ANY_COW
+                if (!(sflags & SVf_IsCOW)) {
+                        SvIsCOW_on(ssv);
+                        CowREFCNT(ssv) = 0;
+                }
+
+                if (len) {
+                   if (sflags & SVf_IsCOW) {
+                        sv_buf_to_rw(ssv);
+                    }
```

Comment:
I'm unclear if there's a bug here carried over from `Perl_sv_setsv_flags`.

If `(!(sflags & SVf_IsCOW))` above, `sflags` isn't updated to reflect that `SVf_Is_COW` is now set on `ssv`, but why would we not want to do `sv_buf_to_rw(ssv)` in that situation?

Reply:
I'm going to make another PR about a very sketchy function in Perl 2003-2025 with 1 parent caller, its name is `sv_setsv_cow()`. I might be mistaken, but your code highlight, and what I think 1 branch in `sv_setsv_cow()` is doing, in my eyes, is a virgin never been HW ROed buffer, that for the first time will be getting HW RO perms added, therefore no ring 0 call is needed to mark it RW perms. verify I am making any sense, this feature isn't available on Windows and I don't see the point/personal necessity to port it to Windows. It would take me 2-5 hours to do that, somewhere between easy and monotonous/unchallenging on the difficulty scale.

Reply:
> I'm unclear if there's a bug here carried over from `Perl_sv_setsv_flags`.
> 
> If `(!(sflags & SVf_IsCOW))` above, `sflags` isn't updated to reflect that `SVf_Is_COW` is now set on `ssv`, but why would we not want to do `sv_buf_to_rw(ssv)` in that situation?

Since this is a carry-over, and one that only applies under debugging, I'm going to resolve this conversation to help clear the way to getting this PR merged early in this dev cycle. (Hopefully will have all current outstanding comments cleared by the end of next weekend.)